lib/autoupdate/agent/updater.go

```diff
@@ -348,18 +360,87 @@ func (u *Updater) Install(ctx context.Context, override OverrideConfig) error {
 	// Only write the configuration file if the initial update succeeds.
 	// Note: skip_version is never set on failed enable, only failed update.
 
-	if err := writeConfig(u.ConfigPath, cfg); err != nil {
+	if err := writeConfig(u.UpdateConfigPath, cfg); err != nil {
 		return trace.Wrap(err, "failed to write %s", updateConfigName)
 	}
 	u.Log.InfoContext(ctx, "Configuration updated.")
 	return trace.Wrap(u.notices(ctx))
 }
 
+// proxyAddrConfig contains potential proxy server addresses from teleport.yaml.
+type proxyAddrConfig struct {
+	Teleport proxyAddrTeleport `yaml:"teleport"`
+}
+
+type proxyAddrTeleport struct {
+	AuthServers []string `yaml:"auth_servers"`
+	AuthServer  string   `yaml:"auth_server"`
+	ProxyServer string   `yaml:"proxy_server"`
+}
+
+// findAgentProxy finds a proxy in teleport.yaml if not specified or set in update configuration.
+// Note that any implicitly defaulted port in teleport.yaml is explicitly defaulted (to 3080) by this method.
+func (u *Updater) findAgentProxy(ctx context.Context) string {
+	f, err := libutils.OpenFileAllowingUnsafeLinks(u.TeleportConfigPath)
+	if err != nil {
+		u.Log.DebugContext(ctx, "Unable to open Teleport config to read proxy", "config", u.TeleportConfigPath, errorKey, err)
+		return ""
+	}
+	defer f.Close()
+	var cfg proxyAddrConfig
+	if err := yaml.NewDecoder(f).Decode(&cfg); err != nil {
+		u.Log.DebugContext(ctx, "Unable to parse Teleport config to read proxy", "config", u.TeleportConfigPath, errorKey, err)
+		return ""
+	}
+
+	var addr string
+	var port int
+	switch t := cfg.Teleport; {
+	case t.ProxyServer != "":
+		addr = t.ProxyServer
+		port = libdefaults.HTTPListenPort
+	case t.AuthServer != "":
+		addr = t.AuthServer
+		port = libdefaults.AuthListenPort
```

Comment:
I don't think `auth_server` ever contains the address of a proxy. `auth_servers` does in v1 and v2.

Reply:
There may be additional validations that prevent this from happening, but it looks like the first auth server address (just `auth_server` in v3) could be used as the resolver address for opening tunnels, which seems like it would allow it to be a proxy address (labeled `ProxyAddr`). Admittedly, I didn't test it. Happy to remove if you think it's not valid.

https://github.com/gravitational/teleport/blob/5590855a2d39910adb1020801fcb244ac512f369/lib/config/configuration.go#L753-L764
https://github.com/gravitational/teleport/blob/5590855a2d39910adb1020801fcb244ac512f369/lib/service/servicecfg/config.go#L411-L413
https://github.com/gravitational/teleport/blob/5590855a2d39910adb1020801fcb244ac512f369/lib/service/service.go#L1222-L1237

Reply:
In the code you linked we use either `proxy_server` or `auth_servers[0]` as the `resolverAddr`, never `auth_server` (without the s).

Reply:
TIL we are doing even more ridiculous logic after

https://github.com/gravitational/teleport/blob/e15bc74885342b12b41945249feeb82162ab0a7d/lib/auth/join/join.go#L287-L310

https://github.com/gravitational/teleport/blob/e15bc74885342b12b41945249feeb82162ab0a7d/lib/auth/join/join.go#L327-L342

I think you're right and it's better to not assume anything about those `proxy_server` and `auth_server`, they were supposed to add semantics but this is apparently not the case 😞

Reply:
If it looks right, waiting on your approval 🙂